src/wasm-binary-reader-objdump.c

```diff
@@ -0,0 +1,434 @@
+  Context* context = ctx->user_data;
+  size_t immediate_len = ctx->offset - context->current_opcode_offset;
+  log_opcode(context, ctx->data, immediate_len, "%d", value);
+  return WASM_OK;
+}
+
+WasmResult on_br_table_expr(WasmBinaryReaderContext* ctx,
+                            uint32_t num_targets,
+                            uint32_t* target_depths,
+                            uint32_t default_target_depth) {
+  Context* context = ctx->user_data;
+  size_t immediate_len = ctx->offset - context->current_opcode_offset;
+  /* TODO(sbc): Print targets */
+  log_opcode(context, ctx->data, immediate_len, NULL);
+  return WASM_OK;
+}
+
+static WasmResult on_end_expr(void* user_data) {
+  Context* context = user_data;
+  context->indent_level--;
+  assert(context->indent_level >= 0);
+  log_opcode(context, NULL, 0, NULL);
+  return WASM_OK;
+}
+
+static const char* wasm_type_name(WasmType type) {
+  switch (type) {
+    case WASM_TYPE_I32:
+      return "i32";
+
+    case WASM_TYPE_I64:
+      return "i64";
+
+    case WASM_TYPE_F32:
+      return "f32";
+
+    case WASM_TYPE_F64:
+      return "f64";
+
+    default:
+      assert(0);
+      return "INVALID TYPE";
+  }
+}
+
+static WasmResult on_opcode_block_sig(WasmBinaryReaderContext* ctx,
+                                      uint32_t num_types,
+                                      WasmType* sig_types) {
+  Context* context = ctx->user_data;
+  if (num_types)
+    log_opcode(context, ctx->data, 1, "%s", wasm_type_name(*sig_types));
+  else
+    log_opcode(context, ctx->data, 1, NULL);
+  context->indent_level++;
+  return WASM_OK;
+}
+
+static WasmResult on_signature(uint32_t index,
+                               uint32_t param_count,
+                               WasmType* param_types,
+                               uint32_t result_count,
+                               WasmType* result_types,
+                               void* user_data) {
+  Context* ctx = user_data;
+
+  if (!ctx->options->verbose)
+    return WASM_OK;
+  printf(" - [%d] (", index);
+  uint32_t i;
+  for (i = 0; i < param_count; i++) {
+    if (i != 0) {
+      printf(", ");
+    }
+    printf("%s", wasm_type_name(param_types[i]));
+  }
+  printf(") -> ");
+  if (result_count)
+    printf("%s", wasm_type_name(result_types[0]));
+  else
+    printf("nil");
+  printf("\n");
+  return WASM_OK;
+}
+
+static WasmResult on_function_signature(uint32_t index,
+                                        uint32_t sig_index,
+                                        void* user_data) {
+  print_details(user_data, " - [%d] sig=%d\n", index, sig_index);
+  return WASM_OK;
+}
+
+static WasmResult begin_function_body(uint32_t index, void* user_data) {
+  Context* ctx = user_data;
+  if (ctx->options->verbose || ctx->options->disassemble)
+    printf("func %d\n", index);
+  ctx->last_opcode_end = 0;
+  return WASM_OK;
+}
+
+static WasmResult on_import_func(uint32_t index,
+                                 uint32_t sig_index,
```

Comment:
Or maybe even better, the combined index space (imports+defined) for globals and functions?